File: internal/relations/subset.go
package relations

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync/atomic"

	"github.com/charmbracelet/log"
	"github.com/videahealth/pg-snap/internal/db"
	"github.com/videahealth/pg-snap/internal/utils"
)

type VisitMode string

const (
	VisitSuccessors   VisitMode = "successors"
	VisitPredecessors VisitMode = "predecessors"
)

type Row struct {
	Table   *db.Table
	ColName string
	Data    string
}

type Subset struct {
	Relations        []db.ForeignKeyInfo
	Tables           []*db.Table
	StartTableName   string
	StartTableSchema string
	SubsetQuery      string
	DAG              DAG
	RootFolder       string
	MaxRowsPerTable  int32
}

func NewSubset(pg *db.Db, tables []*db.Table, startTable, startSchema, rootFolder, query string, maxRowsPerTable int32) (*Subset, error) {
	relations := pg.GetForeignKeys()

	var rels []db.ForeignKeyInfo

	for _, rel := range relations {

		mainTable, _ := GetTable(tables, db.NormalizeName(rel.Schema, rel.Name))
		otherTable, _ := GetTable(tables, db.NormalizeName(rel.ForeignSchema, rel.ForeignName))
		if mainTable != nil && otherTable != nil {
			rels = append(rels, rel)
		}
	}

	var maxRows int32
	if maxRowsPerTable == 0 {
		maxRows = -1
	} else {
		maxRows = maxRowsPerTable
	}

	return &Subset{
		Relations:        rels,
		Tables:           tables,
		StartTableName:   startTable,
		StartTableSchema: startSchema,
		SubsetQuery:      fmt.Sprintf("SELECT * FROM %s WHERE %s", db.NormalizeName(startSchema, startTable), query),
		DAG:              BuildRelations(rels),
		RootFolder:       rootFolder,
		MaxRowsPerTable:  maxRows,
	}, nil
}

// TraverseAndCopyData performs a subset algorithm to copy Table data across a directed acyclic graph (DAG) with the following steps:
//  1. Identifies a new closed system DAG starting from the `startNode`.
//  2. Traverses the graph from the start, generating a list of nodes connected (directly or indirectly) to the `startNode`.
//  3. Iterates over each node, visiting all its predecessors and successors to understand the graph's connectivity.
//  4. Checks if a node's data has already been copied; if so, it skips to the next node to avoid duplication.
//  5. If the node is the `startNode`, it copies its data immediately and proceeds no further with this node.
//  6. For other nodes, it determines the connected nodes based on the traversal direction. If data for these connected nodes has been copied,
//     it resolves the foreign key (FK) relationships. It then copies the data that connects from the current node to the nodes being visited,
//     ensuring data integrity and relational consistency.
func (s *Subset) TraverseAndCopyData() error {

	copiedData := make(map[string]bool)
	gVisited := make(map[string]bool)
	startTable := db.NormalizeName(s.StartTableSchema, s.StartTableName)
	startNode := s.DAG.FindNodeByData(startTable)
	if startNode == nil {
		return fmt.Errorf("table %s not found in database", startTable)
	}
	newDAG := s.DAG.FindClosedSystemFullDAG(startNode)
	var ops atomic.Uint64
	total := len(s.Tables)

	visitNode := func(node *Node, visitMode VisitMode) error {
		table, err := GetTable(s.Tables, node.Data)

		gVisited[node.Data] = true

		if err != nil {
			log.Fatalf("error getting table: %s", err)
			return nil
		}

		isRootNode := node.Data == db.NormalizeName(s.StartTableSchema, s.StartTableName)

		if !copiedData[node.Data] {
			if isRootNode {
				rows := table.PerformCopy(s.RootFolder, s.SubsetQuery)
				utils.DisplayProgress(&ops, rows, total, table.Details.Display)
				copiedData[node.Data] = true
			} else {
				var conditions []string
				var queryNodes []*Node

				switch visitMode {
				case VisitSuccessors:
					queryNodes = node.Children
				case VisitPredecessors:
					queryNodes = newDAG.FindPredecessors(node)
				}
				for _, p := range queryNodes {
					vis := gVisited[p.Data]
					if vis && copiedData[p.Data] {
						var q string
						if visitMode == VisitSuccessors {
							q, err = BuildSuccessorQuery(p.Data, node.Data, s.Relations, s.Tables)
							if err != nil {
								return err
							}
						} else {
							q, err = BuildPredecessorQuery(p.Data, node.Data, s.Relations, s.Tables)
							if err != nil {
								return err
							}
						}
						if q != "" {
							conditions = append(conditions, q)
						}
					}
				}

				if len(conditions) > 0 {
					if node.Data == "\"public\".\"food_des\"" {
						fmt.Println(len(conditions))
					}
					queryCondition := strings.Join(conditions, " OR ")
					var selectSt string
					if s.MaxRowsPerTable == -1 || visitMode == VisitPredecessors {
						selectSt = fmt.Sprintf("SELECT * FROM %s WHERE %s", node.Data, queryCondition)
					} else {
						selectSt = fmt.Sprintf("SELECT * FROM %s WHERE %s LIMIT %d", node.Data, queryCondition, s.MaxRowsPerTable)
					}
					rows := table.PerformCopy(s.RootFolder, selectSt)
					utils.DisplayProgress(&ops, rows, total, table.Details.Display)
					copiedData[node.Data] = true
				}
			}
		}
		return nil
	}

	nodes := newDAG.TraverseGraphFromStart(startNode)
	for {
		totalCopied := 0
		for _, visited := range copiedData {
			if visited {
				totalCopied++
			}
		}
		if totalCopied >= len(newDAG.Nodes) {
			break
		}
		for _, node := range nodes {
			if !copiedData[node.Data] {
				if err := visitNode(node, VisitPredecessors); err != nil {
					return err
				}
				if err := visitNode(node, VisitSuccessors); err != nil {
					return err
				}
			}
		}
	}

	return nil
}

func GetTableDetailsAndCsvPath(tables []*db.Table, toCopyId string, foreignTableId string) (toCopy *db.Table, foreignTable *db.Table, foreignTableCsvPath string, err error) {
	toCopy, err = GetTable(tables, toCopyId)
	if err != nil {
		log.Fatalf("error getting table: %s", err)
	}
	foreignTable, err = GetTable(tables, foreignTableId)
	if err != nil {
		log.Fatalf("error getting table: %s", err)
	}
	foreignTableCsvPath = filepath.Join("./data-dump", foreignTable.Details.Display, "data.csv")
	return
}

func BuildConditions(toCopyId string, foreignTableCsvPath string, cols []db.ForeignKeyInfo, visitMode VisitMode) ([]string, error) {
	var conditions []string

	for _, fk := range cols {
		fkCol := fk.ForeignColumnName
		csvCol := fk.ColumnName
		fkColType := fk.ColType
		if visitMode == VisitSuccessors {
			fkCol = fk.ColumnName
			csvCol = fk.ForeignColumnName
			fkColType = fk.ForeignColType
		}

		data, err := ReadCSVColumnByName(foreignTableCsvPath, csvCol)
		if err != nil {
			log.Fatalf("error getting csv data for table: %s: %s", foreignTableCsvPath, err)
			os.Exit(1)
			continue
		}
		if len(data) == 0 {
			continue
		}

		colData, err := FormatCols(data, fkColType)
		if err != nil {
			return []string{}, err
		}
		if len(colData) == 0 {
			continue
		}
		condition := fmt.Sprintf("\"%s\" IN (%s)", fkCol, colData) // Adjust based on the function requirements
		conditions = append(conditions, condition)
	}

	return conditions, nil
}

func BuildSuccessorQuery(foreignTableId string, toCopyId string, relations []db.ForeignKeyInfo, tables []*db.Table) (string, error) {
	toCopy, foreignTable, foreignTableCsvPath, err := GetTableDetailsAndCsvPath(tables, toCopyId, foreignTableId)
	if err != nil {
		log.Fatalf("error in preparation: %s", err)
	}

	cols := GetTableFk(foreignTable.Details.Schema, foreignTable.Details.Name, toCopy.Details.Schema, toCopy.Details.Name, relations)
	conditions, err := BuildConditions(toCopyId, foreignTableCsvPath, cols, VisitSuccessors)

	if err != nil {
		return "", err
	}

	if len(conditions) != 0 {
		return strings.Join(conditions, " OR "), nil
	}
	return "", nil
}

func BuildPredecessorQuery(foreignTableId string, toCopyId string, relations []db.ForeignKeyInfo, tables []*db.Table) (string, error) {
	toCopy, foreignTable, foreignTableCsvPath, err := GetTableDetailsAndCsvPath(tables, toCopyId, foreignTableId)
	if err != nil {
		log.Fatalf("error in preparation: %s", err)
	}

	cols := GetTableFk(toCopy.Details.Schema, toCopy.Details.Name, foreignTable.Details.Schema, foreignTable.Details.Name, relations)
	conditions, err := BuildConditions(toCopyId, foreignTableCsvPath, cols, VisitPredecessors)
	if err != nil {
		return "", err
	}
	if len(conditions) != 0 {
		return strings.Join(conditions, " OR "), nil
	}
	return "", nil
}

func FormatCols(data []string, colType string) (string, error) {
	var dataVals []string

	category, ok := db.SQLTypeMapping[colType]
	if !ok {
		log.Fatalf("unsupported data type %s", colType)
	}

	for _, d := range data {
		if d == "" {
			continue
		}
		switch {
		case category == "String types" || category == "UUID types" || colType == "varchar" || strings.Contains(category, "Character types"):
			dataVals = append(dataVals, fmt.Sprintf(`'%s'`, d))
		case category == "Numeric types" || category == "Integer types":
			dataVals = append(dataVals, d)
		case category == "Boolean types":
			dataVals = append(dataVals, d)
		case category == "Date/time types":
			dataVals = append(dataVals, fmt.Sprintf(`'%s'`, d))
		case category == "Binary types":
			dataVals = append(dataVals, fmt.Sprintf(`'%s'`, d))
		default:
			return "", fmt.Errorf("unsupported category %s for data type %s", category, colType)
		}
	}

	return strings.Join(dataVals, ","), nil
}

func GetTableFk(ftSchema, ftName, schema, name string, relations []db.ForeignKeyInfo) []db.ForeignKeyInfo {
	var preds []db.ForeignKeyInfo

	for _, rel := range relations {
		if ftSchema == rel.ForeignSchema && ftName == rel.ForeignName && rel.Name == name && rel.Schema == schema {
			preds = append(preds, rel)
		}
	}
	return preds
}

func BuildRelations(relations []db.ForeignKeyInfo) DAG {
	dag := &DAG{}

	for _, rel := range relations {
		node1 := dag.CreateOrGetNode(db.NormalizeName(rel.Schema, rel.Name))
		node2 := dag.CreateOrGetNode(db.NormalizeName(rel.ForeignSchema, rel.ForeignName))

		dag.AddNode(node1)
		dag.AddNode(node2)

		dag.AddEdge(node1, node2)
	}

	return *dag
}

func GetTable(tables []*db.Table, id string) (*db.Table, error) {
	for _, table := range tables {
		if table.Details.Identifier == id {
			return table, nil
		}
	}
	return nil, fmt.Errorf("error table not found for name %s", id)
}

File: src/db_walk/subset.rs
use super::dag::{Node, DAG};
use crate::db::consts::SQL_TYPE_MAPPING;
use crate::db::table::{ForeignKeyInfo, Table};
use crate::db::Db;
use crate::db_walk::copy_data;
use crate::utils::csv::read_csv_column_by_name;
use anyhow::{anyhow, Result};
use colored::Colorize;
use std::collections::HashSet;
use std::path::PathBuf;

#[derive(PartialEq, Debug)]
enum VisitMode {
    Predecessor,
    Successor,
}

pub fn build_relations(relations: Vec<ForeignKeyInfo>) -> DAG {
    let mut dag = DAG::new();

    for rel in relations {
        dag.create_or_get_node(rel.source_table_id.clone());
        dag.create_or_get_node(rel.foreign_table_id.clone());

        dag.add_edge(rel.source_table_id, rel.foreign_table_id);
    }

    dag
}

pub struct Subset {
    relations: Vec<ForeignKeyInfo>,
    tables: Vec<Table>,
    start_table_id: String,
    subset_query: String,
    dag: DAG,
    root_folder: PathBuf,
    max_rows_per_table: Option<i32>,
    db: Db,
    cycles: usize,
}

impl Subset {
    pub fn new(
        tables: Vec<Table>,
        relations: Vec<ForeignKeyInfo>,
        db: Db,
        start_table_name: String,
        start_table_schema: String,
        root_folder: PathBuf,
        max_rows_per_table: Option<i32>,
        subset_query_where: String,
        cycles: usize,
    ) -> Self {
        let dag = build_relations(relations.clone());
        let start_table_id = format!("\"{}\".\"{}\"", start_table_schema, start_table_name);
        let subset_query = format!(
            "SELECT * FROM {} WHERE {}",
            start_table_id, subset_query_where
        );

        Subset {
            relations,
            tables,
            start_table_id,
            db,
            subset_query,
            dag,
            root_folder,
            max_rows_per_table,
            cycles,
        }
    }

    fn build_conditions(
        &self,
        foreign_table_csv_path: &str,
        cols: &[ForeignKeyInfo],
        visit_mode: VisitMode,
    ) -> Result<Vec<String>> {
        let mut conditions: Vec<String> = Vec::new();

        for fk in cols {
            let fk_col = match visit_mode {
                VisitMode::Predecessor => &fk.foreign_column_name,
                VisitMode::Successor => &fk.column_name,
            };
            let csv_col = match visit_mode {
                VisitMode::Predecessor => &fk.column_name,
                VisitMode::Successor => &fk.foreign_column_name,
            };

            let fk_col_type = match visit_mode {
                VisitMode::Predecessor => &fk.col_type,
                VisitMode::Successor => &fk.foreign_col_type,
            };

            let data = read_csv_column_by_name(foreign_table_csv_path, csv_col)?;
            if data.is_empty() {
                continue;
            }

            let col_data = format_cols(data, &fk_col_type)?;

            if col_data.is_empty() {
                continue;
            }

            let condition = format!("\"{}\" IN ({})", fk_col, col_data);
            conditions.push(condition);
        }

        Ok(conditions)
    }

    fn build_successor_query(
        &self,
        foreign_table_id: &str,
        main_table_id: &str,
    ) -> Result<Option<String>> {
        let foreign_table = self
            .tables
            .iter()
            .find(|t| t.id == foreign_table_id)
            .ok_or(anyhow!("Error finding tabe"))?;

        let fks = get_table_fk(foreign_table_id, main_table_id, &self.relations);
        let foreign_table_csv_pathbuf = foreign_table.get_data_dir(&self.root_folder)?;

        let foreign_table_csv_path = foreign_table_csv_pathbuf
            .to_str()
            .ok_or(anyhow!("Unable to get csv path for foreign table"))?;

        let conditions =
            self.build_conditions(foreign_table_csv_path, &fks, VisitMode::Successor)?;

        if !conditions.is_empty() {
            return Ok(Some(conditions.join(" OR ")));
        }

        Ok(None)
    }

    fn build_predecessor_query(
        &self,
        foreign_table_id: &str,
        main_table_id: &str,
    ) -> Result<Option<String>> {
        let foreign_table = self
            .tables
            .iter()
            .find(|t| t.id == foreign_table_id)
            .ok_or(anyhow!("Error finding tabe"))?;

        let fks = get_table_fk(main_table_id, foreign_table_id, &self.relations);
        let foreign_table_csv_pathbuf = foreign_table.get_data_dir(&self.root_folder)?;

        let foreign_table_csv_path = foreign_table_csv_pathbuf
            .to_str()
            .ok_or(anyhow!("Unable to get csv path for foreign table"))?;

        let conditions =
            self.build_conditions(foreign_table_csv_path, &fks, VisitMode::Predecessor)?;

        if !conditions.is_empty() {
            return Ok(Some(conditions.join(" OR ")));
        }

        Ok(None)
    }

    async fn visit_node(
        &self,
        node: &Node,
        visit_mode: VisitMode,
        visited_map: &mut HashSet<String>,
        copied_map: &mut HashSet<String>,
        new_dag: &DAG,
    ) -> Result<()> {
        let visiting_table_id = &node.data;
        let visiting_table = self
            .tables
            .iter()
            .find(|t| t.id == *visiting_table_id)
            .ok_or(anyhow!("Error finding tabe"))?;

        // visited_map.
        visited_map.insert(visiting_table_id.clone());

        let is_root_node = self.start_table_id.eq(visiting_table_id);

        if copied_map.contains(visiting_table_id) {
            return Ok(());
        }

        if is_root_node {
            let rows = copy_data(
                self.db.clone(),
                self.root_folder.clone(),
                visiting_table.clone(),
                Some(&self.subset_query),
            )
            .await?;
            if rows == 0 {
                return Err(anyhow!(
                    "Given subset query {} has no data",
                    self.subset_query.yellow()
                ));
            }
            copied_map.insert(visiting_table_id.clone());
        }

        let mut conditions: Vec<String> = Vec::new();
        let query_nodes = match visit_mode {
            VisitMode::Predecessor => new_dag.find_predecessors(&visiting_table_id),
            VisitMode::Successor => node.children.clone(),
        };

        for query_node in query_nodes {
            let table_id: &String = &query_node.data;
            let visited = visited_map.contains(table_id);
            let copied_data = copied_map.contains(table_id);

            if visited && copied_data {
                let q = match visit_mode {
                    VisitMode::Predecessor => {
                        self.build_predecessor_query(table_id, visiting_table_id)?
                    }
                    VisitMode::Successor => {
                        self.build_successor_query(table_id, visiting_table_id)?
                    }
                };

                if let Some(query) = q {
                    conditions.push(query);
                }
            }
        }

        if conditions.len() == 0 {
            return Ok(());
        }

        let query_condition = conditions.join(" OR ");

        let select_statement = match self.max_rows_per_table {
            Some(max_rows) if visit_mode != VisitMode::Predecessor => {
                format!(
                    "SELECT * FROM {} WHERE {} LIMIT {}",
                    visiting_table_id, query_condition, max_rows
                )
            }
            _ => {
                format!(
                    "SELECT * FROM {} WHERE {}",
                    visiting_table_id, query_condition
                )
            }
        };

        copy_data(
            self.db.clone(),
            self.root_folder.clone(),
            visiting_table.clone(),
            Some(&select_statement),
        )
        .await?;
        copied_map.insert(visiting_table_id.clone());

        Ok(())
    }

    /// Performs a subset algorithm to copy Table data across a directed acyclic graph (DAG) with the following steps:
    ///  1. Identifies a new closed system DAG starting from the `start_table_id`.
    ///  2. Traverses the graph from the start, generating a list of nodes connected (directly or indirectly) to the `start_table_id`.
    ///  3. Iterates over each node, visiting all its predecessors and successors to understand the graph's connectivity.
    ///  4. Checks if a node's data has already been copied; if so, it skips to the next node to avoid duplication.
    ///  5. If the node is the `start_table_id`, it copies its data immediately and proceeds no further with this node.
    ///  6. For other nodes, it determines the connected nodes based on the traversal direction. If data for these connected nodes has been copied.
    ///     it resolves the foreign key (FK) relationships. It then copies the data that connects from the current node to the nodes being visited.
    pub async fn traverse_and_copy_data(&self) -> Result<HashSet<String>> {
        let mut copied_data: HashSet<String> = HashSet::new();
        let mut visited_data: HashSet<String> = HashSet::new();

        let new_dag = self.dag.find_closed_system_full_dag(&self.start_table_id);
        let nodes = new_dag.traverse_graph_from_start(&self.start_table_id);

        let subset_tables: HashSet<_> = nodes
            .clone()
            .into_iter()
            .map(|v| v.data.replace("\"", ""))
            .collect();

        loop {
            let total_copied = &copied_data.len();

            if *total_copied >= self.cycles {
                break;
            }

            for node in &nodes {
                let table_id = &node.data;
                let node = node.clone();

                if !copied_data.contains(table_id) {
                    self.visit_node(
                        &node,
                        VisitMode::Predecessor,
                        &mut visited_data,
                        &mut copied_data,
                        &new_dag,
                    )
                    .await?;
                    self.visit_node(
                        &node,
                        VisitMode::Successor,
                        &mut visited_data,
                        &mut copied_data,
                        &new_dag,
                    )
                    .await?;
                }
            }
        }

        Ok(subset_tables)
    }
}

fn format_cols(data: Vec<String>, col_type: &str) -> Result<String> {
    let mut data_vals: Vec<String> = Vec::new();

    let category = SQL_TYPE_MAPPING
        .get(col_type)
        .ok_or_else(|| anyhow!("unsupported data type {}", col_type))?;

    for d in data {
        if d.is_empty() {
            continue;
        }
        match category {
            &"String types" | &"UUID types" | &"varchar" | &"User-defined types" => {
                data_vals.push(format!("'{}'", d));
            }
            &"Numeric types" | &"Integer types" => {
                data_vals.push(d.to_string());
            }
            &"Boolean types" => {
                data_vals.push(d.to_string());
            }
            &"Date/time types" => {
                data_vals.push(format!("'{}'", d));
            }
            &"Binary types" => {
                data_vals.push(format!("'{}'", d));
            }
            _ => {
                return Err(anyhow!(
                    "unsupported category {} for data type {}",
                    category,
                    col_type
                ))
            }
        }
    }

    Ok(data_vals.join(","))
}

pub fn get_table_fk(
    foreign_table_id: &str,
    main_table_id: &str,
    relations: &[ForeignKeyInfo],
) -> Vec<ForeignKeyInfo> {
    relations
        .iter()
        .filter(|rel| {
            rel.foreign_table_id == foreign_table_id && rel.source_table_id == main_table_id
        })
        .cloned()
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_cols_with_string_types() {
        let data = vec![String::from("Alice"), String::from("Bob")];
        let col_type = "text";
        let formatted = format_cols(data, col_type).unwrap();
        assert_eq!(formatted, "'Alice','Bob'");
    }

    #[test]
    fn test_format_cols_with_numeric_types() {
        let data = vec![String::from("123"), String::from("456")];
        let col_type = "int4";
        let formatted = format_cols(data, col_type).unwrap();
        assert_eq!(formatted, "123,456");
    }

    #[test]
    fn test_format_cols_with_boolean_types() {
        let data = vec![String::from("true"), String::from("false")];
        let col_type = "bool";
        let formatted = format_cols(data, col_type).unwrap();
        assert_eq!(formatted, "true,false");
    }

    #[test]
    fn test_format_cols_with_unsupported_type() {
        let data = vec![String::from("data")];
        let col_type = "UNSUPPORTED_TYPE";
        let result = format_cols(data, col_type);
        assert!(result.is_err());
        assert_eq!(
            result.unwrap_err().to_string(),
            "unsupported data type UNSUPPORTED_TYPE"
        );
    }
}
